uefi/src/proto/media/block.rs

```diff
@@ -186,3 +189,130 @@ impl BlockIOMedia {
+
+    /// Reads the requested number of blocks from the device.
+    ///
+    /// # Arguments
+    /// * `media_id` - The media ID that the read request is for.
+    /// * `lba` - The starting logical block address to read from on the device.
+    /// * `token` - Transaction token for asynchronous read.
+    /// * `len` - Buffer size.
+    /// * `buffer` - The target buffer of the read operation
+    ///
+    /// # Safety
+    /// Because of the asynchronous nature of the block transaction, manual lifetime
+    /// tracking is required.
+    ///
+    /// # Errors
+    /// * [`Status::INVALID_PARAMETER`] The read request contains LBAs that are not valid, or the buffer is not on proper alignment.
+    /// * [`Status::OUT_OF_RESOURCES`]  The request could not be completed due to a lack of resources.
+    /// * [`Status::MEDIA_CHANGED`]     The `media_id` is not for the current media.
+    /// * [`Status::NO_MEDIA`]          There is no media in the device.
+    /// * [`Status::DEVICE_ERROR`]      The device reported an error while performing the read operation.
+    /// * [`Status::BAD_BUFFER_SIZE`]   The buffer size parameter is not a multiple of the intrinsic block size of the device.
+    pub unsafe fn read_blocks_ex(
+        &self,
+        media_id: u32,
+        lba: Lba,
+        token: Option<NonNull<BlockIO2Token>>,
+        len: usize,
+        buffer: *mut u8,
+    ) -> Result {
+        let token = opt_nonnull_to_ptr(token);
+        unsafe { (self.0.read_blocks_ex)(&self.0, media_id, lba, token.cast(), len, buffer.cast()) }
+            .to_result()
+    }
+
+    /// Writes a specified number of blocks to the device.
+    ///
+    /// # Arguments
+    /// * `media_id` - The media ID that the write request is for.
+    /// * `lba` - The starting logical block address to be written.
+    /// * `token` - Transaction token for asynchronous write.
+    /// * `len` - Buffer size.
+    /// * `buffer` - Buffer to be written from.
+    ///
+    /// # Safety
+    /// Because of the asynchronous nature of the block transaction, manual
+    /// lifetime tracking is required.
+    ///
+    /// # Errors
+    /// * [`Status::INVALID_PARAMETER`] The write request contains LBAs that are not valid, or the buffer is not on proper alignment.
+    /// * [`Status::OUT_OF_RESOURCES`]  The request could not be completed due to a lack of resources.
+    /// * [`Status::MEDIA_CHANGED`]     The `media_id` is not for the current media.
+    /// * [`Status::NO_MEDIA`]          There is no media in the device.
+    /// * [`Status::DEVICE_ERROR`]      The device reported an error while performing the write operation.
+    /// * [`Status::WRITE_PROTECTED`]   The device cannot be written to.
+    /// * [`Status::BAD_BUFFER_SIZE`]   The buffer size parameter is not a multiple of the intrinsic block size of the device.
+    pub unsafe fn write_blocks_ex(
+        &mut self,
+        media_id: u32,
+        lba: Lba,
+        token: Option<NonNull<BlockIO2Token>>,
```

Comment:
to me the `Option<NonNull<T>>` looks a little weird and not very idiomatic. However, it is used in lots of places of the code base, so its fine with me.